src/com/amazon/ion/impl/IonReaderContinuableCoreBinary.java

```diff
@@ -0,0 +1,812 @@
+            // new IonException(); However, we choose not to perform that check here because it is not worth sacrificing
+            // performance in this inner-loop code in order to throw one type of exception over another in case of
+            // malformed data.
+            currentByte = buffer[(int) (peekIndex++)];
+            result = (result << VALUE_BITS_PER_VARUINT_BYTE) | (currentByte & LOWER_SEVEN_BITS_BITMASK);
+        } while (currentByte >= 0);
+        return result;
+    }
+
+    /**
+     * Reads a 2+ byte VarInt, given the first byte. When called, `peekIndex` must point at the second byte in the
+     * VarInt. at `peekIndex`.When this method returns, `peekIndex` will point at the first byte that follows the
+     * VarInt. NOTE: the VarInt must fit in an `int`.
+     * @param firstByte the first byte of the VarInt representation, which has already been retrieved from the buffer.
+     * @return the value.
+     */
+    private int readVarInt_1_0(int firstByte) {
+        int currentByte = firstByte;
+        int sign = (currentByte & VAR_INT_SIGN_BITMASK) == 0 ? 1 : -1;
+        int result = currentByte & LOWER_SIX_BITS_BITMASK;
+        while ((currentByte & HIGHEST_BIT_BITMASK) == 0) {
+            // Note: if the varInt is malformed such that it extends beyond the declared length of the value *and*
+            // beyond the end of the buffer, this will result in IndexOutOfBoundsException because only the declared
+            // value length has been filled. Preventing this is simple: if (peekIndex >= valueMarker.endIndex) throw
+            // new IonException(); However, we choose not to perform that check here because it is not worth sacrificing
+            // performance in this inner-loop code in order to throw one type of exception over another in case of
+            // malformed data.
+            currentByte = buffer[(int)(peekIndex++)];
+            result = (result << VALUE_BITS_PER_VARUINT_BYTE) | (currentByte & LOWER_SEVEN_BITS_BITMASK);
+        }
+        return result * sign;
+    }
+
+    /**
+     * Reads the VarInt starting at `peekIndex`. When this method returns, `peekIndex` will point at the first byte
+     * that follows the VarUInt. NOTE: the VarInt must fit in an `int`.
+     * @return the value.
+     */
+    private int readVarInt_1_0() {
+        return readVarInt_1_0(buffer[(int)(peekIndex++)]);
+    }
+
+    /**
+     * Reads into a BigInteger the UInt value that begins at `valueMarker.startIndex` and ends at
+     * `valueMarker.endIndex`.
+     * @param isNegative true if the resulting BigInteger value should be negative; false if it should be positive.
+     * @return the value.
+     */
+    private BigInteger readUIntAsBigInteger(boolean isNegative) {
+        int length = (int) (valueMarker.endIndex - valueMarker.startIndex);
+        // NOTE: unfortunately, there is no BigInteger(int signum, byte[] bits, int offset, int length) constructor
```

Comment:
That technique will just give us back our buffer as-is, because wrapping with ByteBuffer uses the given array directly. Unfortunately, we have to copy to a right-sized array to comply with the BigInteger constructor. As of JDK 9 we will be able to simply do `new BigInteger(buffer, valueMarker.startIndex, valueMarker.endIndex - valueMarker.startIndex)`